src/contracts/middleware/IndexRegistry.sol

```diff
@@ -0,0 +1,263 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
+
+
+import "../interfaces/IIndexRegistry.sol";
+import "../interfaces/IRegistryCoordinator.sol";
+import "../libraries/BN254.sol";
+
+contract IndexRegistry is IIndexRegistry {
+
+    /// @notice The value that indices of deregistered operators are set to
+    uint32 public constant OPERATOR_DEREGISTERED_INDEX = type(uint32).max;
+
+    IRegistryCoordinator public immutable registryCoordinator;
+
+    // list of all operators ever registered, may include duplicates. used to avoid running an indexer on nodes
+    bytes32[] public globalOperatorList;
+
+    // mapping of operatorId => quorumNumber => index history of that operator
+    mapping(bytes32 => mapping(uint8 => OperatorIndexUpdate[])) internal _operatorIdToIndexHistory;
+    // mapping of quorumNumber => history of numbers of unique registered operators
+    mapping(uint8 => OperatorIndexUpdate[]) internal _totalOperatorsHistory;
+
+    modifier onlyRegistryCoordinator() {
+        require(msg.sender == address(registryCoordinator), "IndexRegistry.onlyRegistryCoordinator: caller is not the registry coordinator");
+        _;
+    }
+
+    constructor(
+        IRegistryCoordinator _registryCoordinator
+    ){
+        registryCoordinator = _registryCoordinator;
+    }
+
```

Comment:
similar question to BLSPubkeyRegistry -- is this contract supposed to be upgradeable or not?

Reply:
confirmed: this should be upgradeable.

Reply:
this was resolved with PR https://github.com/Layr-Labs/eigenlayer-contracts/pull/216